### PR TITLE
Provide APIs to read file into more data types

### PR DESCRIPTION
**Pull request by clintonpi**
Motivation:

As requested in issues [#2875](https://github.com/apple/swift-nio/issues/2875) and [#2876](https://github.com/apple/swift-nio/issues/2876), it would be convenient to be able to read the contents of a file into more data types such as `Array`, `ArraySlice` & Foundation's `Data`.

Modifications:

- Extend `Array`, `ArraySlice` & `Data` to be initialisable with the contents of a file.

Result:

The contents of a file can be read into more data types.

### CODE REVIEW

File: Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        fileSystem: some FileSystemProtocol
+    ) async throws {
+        let byteBuffer = try await fileSystem.withFileHandle(forReadingAt: path) { handle in
+            try await handle.readToEnd(
+                fromAbsoluteOffset: 0,
+                maximumSizeAllowed: maximumSizeAllowed
+            )
+        }
+
+        self = byteBuffer.withUnsafeReadableBytes { Self($0) }
```
Comment (glbrntt) [inline Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift:40]: The [NIOFoundationCompat](https://github.com/apple/swift-nio/tree/main/Sources/NIOFoundationCompat) module provides an extension on `Data` which you should use instead: https://github.com/apple/swift-nio/blob/f7dc3f527576c398709b017584392fb58592e7f5/Sources/NIOFoundationCompat/ByteBuffer-foundation.swift#L403-L406

File: Sources/NIOFileSystem/Array+FileSystem.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public init(
+        contentsOf path: FilePath,
+        maximumSizeAllowed: ByteCount,
+        fileSystem: some FileSystemProtocol
+    ) async throws {
+        let byteBuffer = try await fileSystem.withFileHandle(forReadingAt: path) { handle in
+            try await handle.readToEnd(
+                fromAbsoluteOffset: 0,
+                maximumSizeAllowed: maximumSizeAllowed
+            )
```
Comment (glbrntt) [inline Sources/NIOFileSystem/Array+FileSystem.swift:35]: 0 is the default, so no need to specify it here
```suggestion
            try await handle.readToEnd(maximumSizeAllowed: maximumSizeAllowed)
```

File: Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    public init(
+        contentsOf path: FilePath,
+        maximumSizeAllowed: ByteCount,
+        fileSystem: some FileSystemProtocol
+    ) async throws {
+        let byteBuffer = try await fileSystem.withFileHandle(forReadingAt: path) { handle in
+            try await handle.readToEnd(
+                fromAbsoluteOffset: 0,
+                maximumSizeAllowed: maximumSizeAllowed
+            )
```
Comment (glbrntt) [inline Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift:37]: Same here re: 0
```suggestion
            try await handle.readToEnd(maximumSizeAllowed: maximumSizeAllowed)
```

File: Tests/NIOFileSystemFoundationCompatTests/FileSystemFoundationCompatTests.swift [hunk 4]
```diff
@@ -33,5 +33,18 @@ final class FileSystemBytesConformanceTests: XCTestCase {
             Date(timeIntervalSince1970: 1.000000001)
         )
     }
+
+    func testReadFileIntoData() async throws {
+        let fs = FileSystem.shared
+        let path = FilePath(#filePath)
+
+        try await fs.withFileHandle(forReadingAndWritingAt: path) { fileHandle in
```
Comment (glbrntt) [inline Tests/NIOFileSystemFoundationCompatTests/FileSystemFoundationCompatTests.swift:41]: Writing into this source file is a _bad_ idea (think about the local development experience of running this test). Reading is fine, writing less so.

File: Tests/NIOFileSystemIntegrationTests/FileSystemTests.swift [hunk 5]
```diff
@@ -1828,6 +1828,30 @@ extension FileSystemTests {
...
+        let path = FilePath(#filePath)
+
+        try await self.fs.withFileHandle(forReadingAndWritingAt: path) { fileHandle in
+            _ = try await fileHandle.write(contentsOf: [0, 1, 2], toAbsoluteOffset: 0)
+        }
+
+        let contents = try await ArraySlice(contentsOf: path, maximumSizeAllowed: .bytes(1024))
+
+        XCTAssertEqual(contents, [0, 1, 2])
+    }
```
Comment (glbrntt) [inline Tests/NIOFileSystemIntegrationTests/FileSystemTests.swift:1854]: Same comments apply here as well: please don't write into this source file.

**Review by glbrntt — CHANGES_REQUESTED** [review]
Looks good but a few things need fixing up before we can merge this.

File: Tests/NIOFileSystemFoundationCompatTests/FileSystemFoundationCompatTests.swift (hunk 4 above)
Comment (clintonpi) [reply Tests/NIOFileSystemFoundationCompatTests/FileSystemFoundationCompatTests.swift:41]: True, I missed that. I worked with `temporaryFilePath()` locally but swapped it with `FilePath(#filePath)` just before committing. `FilePath(#filePath)` fails for me locally due to permission issues so I was depending on the CI tests and didn't re-test locally after swapping.

File: Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    ///
+    /// - Parameters:
+    ///   - path: The path of the file to read.
+    ///   - maximumSizeAllowed: The maximum size of file which can be read, in bytes, as a ``ByteCount``.
+    ///   - fileSystem: The ``FileSystemProtocol`` instance to use to read the file.
+    public init(
+        contentsOf path: FilePath,
+        maximumSizeAllowed: ByteCount,
+        fileSystem: some FileSystemProtocol,
+        byteTransferStrategy: ByteBuffer.ByteTransferStrategy = .automatic
```
Comment (glbrntt) [inline Sources/NIOFileSystemFoundationCompat/Data+FileSystem.swift:33]: We don't think we need to expose this, we can just use the default behaviour

**Review by glbrntt — CHANGES_REQUESTED** [review]
One nit, but looks good otherwise.

**Pull request merged**